shim/opentelemetry-opencensus-shim/src/opentelemetry/shim/opencensus/_shim_span.py

```diff
@@ -0,0 +1,146 @@
+
+_MESSAGE_EVENT_TYPE_STR_MAPPING = {
+    0: "TYPE_UNSPECIFIED",
+    1: "SENT",
+    2: "RECEIVED",
+}
+
+
+# pylint: disable=abstract-method
+class ShimSpan(wrapt.ObjectProxy):
+    def __init__(
+        self,
+        wrapped: BaseSpan,
+        *,
+        otel_span: trace.Span,
+        shim_tracer: "ShimTracer",
+    ) -> None:
+        super().__init__(wrapped)
+        self._self_otel_span = otel_span
+        self._self_shim_tracer = shim_tracer
+        self._self_token: object = None
+
+        # Set a few values for BlankSpan members (they appear to be part of the "public" API
+        # even though they are not documented in BaseSpan). Some instrumentations may use these
+        # and not expect an AttributeError to be raised. Set values from OTel where possible
+        # and let ObjectProxy defer to the wrapped BlankSpan otherwise.
+        sc = self._self_otel_span.get_span_context()
+        self.same_process_as_parent_span = not sc.is_remote
+        self.span_id = sc.span_id
+
+    def span(self, name="child_span"):
+        return self._self_shim_tracer.start_span(name=name)
+
+    def add_attribute(self, attribute_key, attribute_value):
+        self._self_otel_span.set_attribute(attribute_key, attribute_value)
+
+    def add_annotation(self, description, **attrs):
+        self._self_otel_span.add_event(description, attrs)
+
+    def add_message_event(self, message_event: MessageEvent):
+        attrs = {
+            _MESSAGE_EVENT_ATTRIBUTE_KEY_TYPE: _MESSAGE_EVENT_TYPE_STR_MAPPING[
+                message_event.type
+            ],
+        }
+        if message_event.uncompressed_size_bytes is not None:
+            attrs[
+                _MESSAGE_EVENT_ATTRIBUTE_KEY_SIZE_UNCOMPRESSED
+            ] = message_event.uncompressed_size_bytes
+        if message_event.compressed_size_bytes is not None:
+            attrs[
+                _MESSAGE_EVENT_ATTRIBUTE_KEY_SIZE_COMPRESSED
+            ] = message_event.compressed_size_bytes
+
+        self._self_otel_span.add_event(
+            str(message_event.id),
+            attrs,
+        )
+
+    # pylint: disable=no-self-use
+    def add_link(self, link):
+        """span links do not work with the shim because the OpenCensus Tracer does not accept
+        links in start_span(). Same issue applies to SpanKind. Also see:
+        https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/compatibility/opencensus.md#known-incompatibilities
+        """
+        _logger.warning(
+            "OpenTelemetry does not support links added after a span is created."
+        )
+
+    @property
+    def span_kind(self):
+        """Setting span_kind does not work with the shim because the OpenCensus Tracer does not
+        accept the param in start_span() and there's no way to set OTel span kind after
+        start_span().
+        """
+        return SpanKind.UNSPECIFIED
```

Comment:
I'm wondering how we can help the user differentiate which properties are unsupported due to the difference in apis between OT and OC (like spankind and add_links) and properties that just can't be found in OT api (like stack_trace, context_tracer, attributes, annotations, etc.) because we don't have getters.

Reply:
Sure what do you have in mind? Just docs?

I added log warnings in those areas where things will be dropped (might make them just INFO level). See e.g. L117